src/data_race.rs

```diff
@@ -0,0 +1,1189 @@
+///  of a thread, contains the happens-before clock and
+///  additional metadata to model atomic fence operations
+#[derive(Clone, Default, Debug)]
+struct ThreadClockSet {
+
+    /// The increasing clock representing timestamps
+    ///  that happen-before this thread.
+    clock: VClock,
+
+    /// The set of timestamps that will happen-before this
+    ///  thread once it performs an acquire fence
+    fence_acquire: VClock,
+
+    /// The last timesamp of happens-before relations that
+    ///  have been released by this thread by a fence
+    fence_release: VClock,
+}
+
+impl ThreadClockSet {
+
+    /// Apply the effects of a release fence to this
+    ///  set of thread vector clocks
+    #[inline]
+    fn apply_release_fence(&mut self) {
+        self.fence_release.clone_from(&self.clock);
+    }
+
+    /// Apply the effects of a acquire fence to this
+    ///  set of thread vector clocks
+    #[inline]
+    fn apply_acquire_fence(&mut self) {
+        self.clock.join(&self.fence_acquire);
+    }
+
+    /// Increment the happens-before clock at a
+    ///  known index
+    #[inline]
+    fn increment_clock(&mut self, index: VectorIdx) {
+        self.clock.increment_index(index);
+    }
+
+    /// Join the happens-before clock with that of
+    ///  another thread, used to model thread join
+    ///  operations
+    fn join_with(&mut self, other: &ThreadClockSet) {
+        self.clock.join(&other.clock);
+    }
+}
+
+/// Extra metadata associated with a thread
+#[derive(Debug, Clone, Default)]
+struct ThreadExtraState {
+
+    /// The current vector index in use by the
+    ///  thread currently, this is set to None
+    ///  after the vector index has been re-used
+    ///  and hence the value will never need to be
+    ///  read during data-race reporting
+    vector_index: Option<VectorIdx>,
+
+    /// The name of the thread, updated for better
+    ///  diagnostics when reporting detected data
+    ///  races
+    thread_name: Option<Box<str>>,
+    
+    /// Thread termination vector clock, this
+    ///  is set on thread termination and is used
+    ///  for joining on threads since the vector_index
+    ///  may be re-used when the join operation occurs
+    termination_vector_clock: Option<VClock>,
+}
+
+/// Global data-race detection state, contains the currently
+///  executing thread as well as the vector-clocks associated
+///  with each of the threads.
+#[derive(Debug, Clone)]
+pub struct GlobalState {
+
+    /// Set to true once the first additional
+    ///  thread has launched, due to the dependency
+    ///  between before and after a thread launch
+    /// Any data-races must be recorded after this
+    ///  so concurrent execution can ignore recording
+    ///  any data-races
+    multi_threaded: Cell<bool>,
+
+    /// Mapping of a vector index to a known set of thread
+    ///  clocks, this is not directly mapping from a thread id
+    ///  since it may refer to multiple threads
```

Comment:
Can you add a top-level comment somewhere explaining this entire business about reusing `VectorIdx` and one idx being for multiple threads? It seems very non-trivial.